pisa/utils/hypersurface.py

```diff
@@ -0,0 +1,1927 @@
+        sys_param_values : list of dicts
+            List where each element if a dict containing the values of each systematic param used to generate the that dataset
+            Each list element specified the parameters for the corresponding element in `sys_maps`
+
+        norm : bool
+            Normalise the maps to the nominal map.
+            This is what you want to do when using the hypersurface to re-weight simulation (which is the main use case).
+            In principal the hypersurfaces are more general though and could be used for other tasks too, hence this option.
+
+        method : str
+            `method` arg to pass to `scipy.optimize.curve_fit`
+
+        smooth : str
+            Smoothing method to use. Choose `None` if do not want smoothing.
+            Smoothing methods supported:
+                - `gaussian_filter`
+
+        smooth_kw : dict
+            kwargs to pass to smoothing method underlying function
+            Format depends on smoothing method:
+              `gaussian_filter`
+                 kwargs for `scipy.ndimage.filters.gaussian_filter`
+                 MUST include `sigma`, `order`
+        '''
+
+        #TODO Add option to exclude bins with too few stats from the fit, leving null hypersurface for them.
+        #     This is to avoid issues with bins with tiny stats having crazy gradients from statistical 
+        #     fluctuations (if there are very few events in that bin for that speciesi then that bin shouldn't
+        #     be significant in the fit).
+
+
+        #
+        # Check inputs
+        #
+
+        # Check nominal dataset definition
+        assert isinstance(nominal_map, Map)
+        assert isinstance(nominal_param_values, collections.Mapping)
+        assert set(nominal_param_values.keys()) == set(self.param_names)
+        assert all([ isinstance(k, str) for k in nominal_param_values.keys() ])
+        assert all([ np.isscalar(v) for v in nominal_param_values.values() ])
+
+        # Check systematic dataset definitions
+        assert isinstance(sys_maps, collections.Sequence)
+        assert isinstance(sys_param_values, collections.Sequence)
+        assert len(sys_maps) == len(sys_param_values)
+        for sys_map, sys_param_vals in zip(sys_maps, sys_param_values) :
+            assert isinstance(sys_map, Map)
+            assert isinstance(sys_param_vals, collections.Mapping)
+            assert set(sys_param_vals.keys()) == set(self.param_names)
+            assert all([ isinstance(k, str) for k in sys_param_vals.keys() ])
+            assert all([ np.isscalar(v) for v in sys_param_vals.values() ])
+            assert sys_map.binning == nominal_map.binning
+
+
+        #
+        # Format things before getting started
+        #
+
+        # Default fit method
+        # Choosing one that produces covariance matrix results reliably
+        self.fit_method = method
+        if self.fit_method is None :
+            self.fit_method = "lm"  # lm, trf, dogbox
```

Comment:
In general I prefer to set default args values in this manner because otherwise if a higher level function calls this function and itself takes `fit_method` as an arg then it also may have to reproduce the same default value. Same situation if the argument is set via e.g. argparse. The defaults values can then diverge. 

This way the default value is always `None` and only in a single place do you need to define the default value. 

But do not care enough about this if you have a strong preference for the alternative?

Reply:
I don't understand what you mean by this. If you use `argparse`, then if a user doesn't specify an arg at all (assuming you don't define a `default` for that arg in your parser, which you already said you didn't want to do), it will not be produced in the resulting `namespace` that `parse_args()` produces at all and therefore calling the function via `func(**vars(parsed))`, the variable will not be present and hence the default will be set from the function definition, e.g. `default_x_val` in `def func(x=default_x_val)`.

See docs here: https://docs.python.org/3/library/argparse.html and tutorial here, with link to relevant section here: https://docs.python.org/3/howto/argparse.html#introducing-optional-arguments

I have a strong preference (as you should as well) to use the features of the language to make your code more readable, explicit, and concise.

It seems you have a pattern of re-creating many things that Python, Numpy, etc. already do. Maybe because you don't know those features exist, or you don't understand how those features in the language work and, rather than read the docs on them or figure out what's there, you write your own code to do the same thing. This is of course error-prone, takes more time for you, makes your code less readable, and harder to maintain.

Just general advice for improving your code, and hopefully saving you time in the long run (and heping others who review / use / maintain your code).